File: pdf_extractor.py
import os
import fitz
from datetime import datetime

def parse_pdf_date(date_str):
    """
    Parses a PDF date string (e.g. 'D:20260520220137Z' or 'D:20260520220137+05\'30\'')
    into a clean, human-readable format: YYYY-MM-DD HH:MM:SS.
    """
    if not date_str:
        return ""
    
    clean_date = date_str
    if clean_date.startswith("D:"):
        clean_date = clean_date[2:]
        
    # Replace single quotes in timezone offset (e.g. +05'30' -> +0530)
    clean_date = clean_date.replace("'", "")
    
    try:
        if len(clean_date) >= 14:
            year = clean_date[0:4]
            month = clean_date[4:6]
            day = clean_date[6:8]
            hour = clean_date[8:10]
            minute = clean_date[10:12]
            second = clean_date[12:14]
            
            tz = ""
            if len(clean_date) > 14:
                # e.g., +0530 or -0800 or Z
                tz_part = clean_date[14:]
                if tz_part == "Z":
                    tz = " UTC"
                else:
                    # format offset as e.g., +05:30
                    if len(tz_part) >= 5 and (tz_part[0] in ('+', '-')):
                        tz = f" (GMT{tz_part[0:3]}:{tz_part[3:5]})"
                    else:
                        tz = f" ({tz_part})"
            return f"{year}-{month}-{day} {hour}:{minute}:{second}{tz}"
        elif len(clean_date) >= 8:
            year = clean_date[0:4]
            month = clean_date[4:6]
            day = clean_date[6:8]
            return f"{year}-{month}-{day}"
    except Exception:
        pass
    
    return date_str

def get_annotation_type_name(annot_type_info):
    """
    Translates PyMuPDF annotation type representations to a clean string.
    Works across different PyMuPDF versions.
    """
    if isinstance(annot_type_info, tuple) and len(annot_type_info) == 2:
        return annot_type_info[1]
    
    if isinstance(annot_type_info, int):
        # Standard PDF Annotation Type Numbers mapping
        mapping = {
            0: 'Text',       # Sticky Note
            1: 'Link',
            2: 'FreeText',   # TextBox
            3: 'Line',
            4: 'Square',     # Rectangle
            5: 'Circle',
            6: 'Polygon',
            7: 'PolyLine',
            8: 'Highlight',
            9: 'Underline',
            10: 'Squiggly',
            11: 'StrikeOut',
            12: 'Stamp',
            13: 'Caret',
            14: 'Ink',       # Freehand drawing
            15: 'Popup',
            16: 'FileAttachment',
            17: 'Sound',
            18: 'Movie',
            19: 'Widget',
            20: 'Screen',
            21: 'PrinterMark',
            22: 'TrapNet',
            23: 'Watermark',
            24: '3D',
            25: 'Redact'
        }
        return mapping.get(annot_type_info, f"Unknown ({annot_type_info})")
    
    return str(annot_type_info)

def extract_annotations(pdf_path, pdf_name=None):
    """
    Opens a PDF and extracts all annotations/markups.
    Returns a list of dicts.
    """
    if not pdf_name:
        pdf_name = os.path.basename(pdf_path)
        
    extracted_markups = []
    
    try:
        doc = fitz.open(pdf_path)
    except Exception as e:
        raise ValueError(f"Failed to open PDF file: {str(e)}")
        
    for page_num in range(len(doc)):
        page = doc[page_num]
        
        # page.annots() returns a generator of annotations on this page
        annots = page.annots()
        if not annots:
            continue
            
        for annot in annots:
            # Skip link annotations or widgets that are not markups
            # We map type first
            annot_type_info = annot.type
            annot_type = get_annotation_type_name(annot_type_info)
            
            if annot_type in ('Link', 'Widget'):
                continue
                
            info = annot.info
            rect = annot.rect
            
            # Extract coordinates as string [x0, y0, x1, y1]
            coords_str = ""
            if rect:
                coords_str = f"[{round(rect.x0, 1)}, {round(rect.y0, 1)}, {round(rect.x1, 1)}, {round(rect.y1, 1)}]"
                
            # Extract text comment/contents
            comment_text = info.get("content", "").strip()
            
            # If there's no comment text, use a placeholder based on type
            if not comment_text:
                comment_text = f"[{annot_type} Markup - No comment text]"
                
            # Extract author (title field in info is usually author name)
            author = info.get("title", "").strip()
            if not author:
                author = "Unknown Reviewer"
                
            # Extract and parse dates
            created_date_raw = info.get("creationDate", "")
            modified_date_raw = info.get("modDate", "")
            
            created_date = parse_pdf_date(created_date_raw)
            modified_date = parse_pdf_date(modified_date_raw)
            
            # If created date is empty, fallback to modification date or current time
            if not created_date:
                if modified_date:
                    created_date = modified_date
                else:
                    created_date = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            if not modified_date:
                modified_date = created_date
                
            # Extract underlying selected text if annotation is over readable text (e.g. highlight)
            selected_text = ""
            if rect and annot_type in ('Highlight', 'Underline', 'StrikeOut', 'Squiggly'):
                try:
                    selected_text = page.get_text("text", clip=rect).strip()
                except Exception:
                    pass
            
            extracted_markups.append({
                "pdf_name": pdf_name,
                "page_number": page_num + 1,  # 1-indexed for display
                "annotation_type": annot_type,
                "comment_text": comment_text,
                "author": author,
                "created_date": created_date,
                "modified_date": modified_date,
                "rectangle_coordinates": coords_str,
                "selected_text": selected_text,
                "assigned_to": "Unassigned",
                "priority": "Medium",
                "status": "Pending",
                "remarks": ""
            })
            
    doc.close()
    return extracted_markups

File: lib/pdf-extractor.ts
/**
 * PDF Annotation Extractor — JavaScript implementation
 * Mirrors the Python PyMuPDF (fitz) extractor logic using pdf.js
 */

import type { ExtractedAnnotation } from './types';

// PDF annotation type number → name mapping (matches PDF spec & Python extractor)
const ANNOT_TYPE_MAP: Record<number, string> = {
  1: 'Text',
  2: 'Link',
  3: 'FreeText',
  4: 'Line',
  5: 'Square',
  6: 'Circle',
  7: 'Polygon',
  8: 'PolyLine',
  9: 'Highlight',
  10: 'Underline',
  11: 'Squiggly',
  12: 'StrikeOut',
  13: 'Stamp',
  14: 'Caret',
  15: 'Ink',
  16: 'Popup',
  17: 'FileAttachment',
  18: 'Sound',
  19: 'Movie',
  20: 'Widget',
  21: 'Screen',
  22: 'PrinterMark',
  23: 'TrapNet',
  24: 'Watermark',
  25: '3D',
  26: 'Redact',
};

const SKIP_TYPES = new Set(['Link', 'Widget', 'Popup']);
const TEXT_EXTRACT_TYPES = new Set(['Highlight', 'Underline', 'StrikeOut', 'Squiggly']);

/**
 * Parse a PDF date string like "D:20260520220137Z" into a readable format.
 */
function parsePdfDate(dateStr: string): string {
  if (!dateStr) return '';

  let clean = dateStr;
  if (clean.startsWith('D:')) clean = clean.slice(2);
  clean = clean.replace(/'/g, '');

  try {
    if (clean.length >= 14) {
      const year   = clean.slice(0, 4);
      const month  = clean.slice(4, 6);
      const day    = clean.slice(6, 8);
      const hour   = clean.slice(8, 10);
      const min    = clean.slice(10, 12);
      const sec    = clean.slice(12, 14);
      let tz = '';
      if (clean.length > 14) {
        const tzPart = clean.slice(14);
        if (tzPart === 'Z') {
          tz = ' UTC';
        } else if (tzPart.length >= 5 && (tzPart[0] === '+' || tzPart[0] === '-')) {
          tz = ` (GMT${tzPart.slice(0, 3)}:${tzPart.slice(3, 5)})`;
        } else {
          tz = ` (${tzPart})`;
        }
      }
      return `${year}-${month}-${day} ${hour}:${min}:${sec}${tz}`;
    }
    if (clean.length >= 8) {
      return `${clean.slice(0, 4)}-${clean.slice(4, 6)}-${clean.slice(6, 8)}`;
    }
  } catch {
    // fall through
  }
  return dateStr;
}

function getAnnotTypeName(subtype: string | number | undefined): string {
  if (typeof subtype === 'string') {
    return subtype || 'Unknown';
  }
  if (typeof subtype === 'number') {
    return ANNOT_TYPE_MAP[subtype] ?? `Unknown (${subtype})`;
  }
  return 'Unknown';
}

function nowFormatted(): string {
  return new Date().toISOString().replace('T', ' ').slice(0, 19);
}

function getAnnotationColor(colorArray?: any): string {
  if (!colorArray || (colorArray.length !== 3 && colorArray.length !== 4)) {
    return 'Yellow'; // Default fallback
  }

  // Handle float arrays where elements are 0-1
  const r = colorArray[0] <= 1 && colorArray[1] <= 1 && colorArray[2] <= 1
    ? Math.round(colorArray[0] * 255)
    : Math.round(colorArray[0]);
  const g = colorArray[1] <= 1 && colorArray[1] <= 1 && colorArray[2] <= 1
    ? Math.round(colorArray[1] * 255)
    : Math.round(colorArray[1]);
  const b = colorArray[2] <= 1 && colorArray[1] <= 1 && colorArray[2] <= 1
    ? Math.round(colorArray[2] * 255)
    : Math.round(colorArray[2]);

  // Curated color palette mapping closest Euclidean distance in RGB
  const targets = [
    { name: 'Yellow', r: 250, g: 220, b: 80 },
    { name: 'Green',  r: 100, g: 220, b: 120 },
    { name: 'Blue',   r: 100, g: 200, b: 250 },
    { name: 'Pink',   r: 250, g: 120, b: 180 },
    { name: 'Orange', r: 250, g: 160, b: 80 },
    { name: 'Purple', r: 180, g: 140, b: 250 },
  ];

  let closest = targets[0];
  let minDistance = Infinity;

  for (const t of targets) {
    const dist = Math.sqrt(
      Math.pow(r - t.r, 2) +
      Math.pow(g - t.g, 2) +
      Math.pow(b - t.b, 2)
    );
    if (dist < minDistance) {
      minDistance = dist;
      closest = t;
    }
  }

  return closest.name;
}

/**
 * Polyfill DOMMatrix for Node.js / Vercel serverless environments.
 * pdfjs-dist internally references DOMMatrix which only exists in browsers.
 */
function ensureDOMMatrixPolyfill() {
  if (typeof globalThis.DOMMatrix === 'undefined') {
    // Minimal stub — pdfjs uses it for coordinate transforms but we only
    // need annotation metadata, not rendering, so a no-op is fine.
    (globalThis as any).DOMMatrix = class DOMMatrix {
      a = 1; b = 0; c = 0; d = 1; e = 0; f = 0;
      m11 = 1; m12 = 0; m13 = 0; m14 = 0;
      m21 = 0; m22 = 1; m23 = 0; m24 = 0;
      m31 = 0; m32 = 0; m33 = 1; m34 = 0;
      m41 = 0; m42 = 0; m43 = 0; m44 = 1;
      is2D = true; isIdentity = true;

      constructor(init?: any) {
        if (Array.isArray(init) && init.length === 6) {
          [this.a, this.b, this.c, this.d, this.e, this.f] = init;
          this.m11 = this.a; this.m12 = this.b;
          this.m21 = this.c; this.m22 = this.d;
          this.m41 = this.e; this.m42 = this.f;
        }
      }

      inverse() { return new DOMMatrix(); }
      multiply() { return new DOMMatrix(); }
      translate() { return new DOMMatrix(); }
      scale() { return new DOMMatrix(); }
      rotate() { return new DOMMatrix(); }
      transformPoint(p: any) { return p || { x: 0, y: 0, z: 0, w: 1 }; }
      static fromMatrix() { return new DOMMatrix(); }
      static fromFloat32Array() { return new DOMMatrix(); }
      static fromFloat64Array() { return new DOMMatrix(); }
    };
  }

  // Also stub Path2D if missing (pdfjs may reference it)
  if (typeof globalThis.Path2D === 'undefined') {
    (globalThis as any).Path2D = class Path2D {
      constructor(_d?: string | Path2D) {}
      addPath() {}
      closePath() {}
      moveTo() {}
      lineTo() {}
      bezierCurveTo() {}
      quadraticCurveTo() {}
      arc() {}
      arcTo() {}
      ellipse() {}
      rect() {}
    };
  }
}

async function ensurePdfWorker() {
  if ((globalThis as any).pdfjsWorker?.WorkerMessageHandler) {
    return;
  }

  const workerModule = await import('pdfjs-dist/legacy/build/pdf.worker.mjs');
  (globalThis as any).pdfjsWorker = workerModule;
}

/**
 * Extract annotations from a PDF buffer using pdf.js
 * Returns a flat list of ExtractedAnnotation ready for DB insert.
 */
export async function extractAnnotations(
  buffer: ArrayBuffer,
  pdfName: string,
  pdfUrl: string = ''
): Promise<ExtractedAnnotation[]> {
  // Polyfill browser globals before importing pdfjs-dist
  ensureDOMMatrixPolyfill();

  // Dynamic import — pdf.js is an ESM package
  const pdfjsLib = await import('pdfjs-dist/legacy/build/pdf.mjs').catch(async () => {
    return await import('pdfjs-dist');
  });

  const pdfjs = (pdfjsLib as any).default ?? pdfjsLib;

  await ensurePdfWorker();

  const loadingTask = pdfjs.getDocument({
    data: new Uint8Array(buffer),
    useWorkerFetch: false,
    isEvalSupported: false,
    disableFontFace: true,
    useSystemFonts: false,
  });
  const doc = await loadingTask.promise;

  const results: ExtractedAnnotation[] = [];
  const numPages = doc.numPages;

  for (let pageNum = 1; pageNum <= numPages; pageNum++) {
    const page = await doc.getPage(pageNum);
    const annotations = await page.getAnnotations({ intent: 'display' });

    for (const annot of annotations) {
      const subtypeName = getAnnotTypeName(annot.subtype);

      if (SKIP_TYPES.has(subtypeName)) continue;

      // Coordinates: [x0, y0, x1, y1] — pdf.js gives rect as [x0, y0, x1, y1]
      let coordsStr = '';
      if (annot.rect && Array.isArray(annot.rect) && annot.rect.length === 4) {
        const [x0, y0, x1, y1] = annot.rect.map((v: number) => Math.round(v * 10) / 10);
        coordsStr = `[${x0}, ${y0}, ${x1}, ${y1}]`;
      }

      // Comment text (contents field)
      let commentText = (annot.contents ?? '').trim();
      if (!commentText) {
        commentText = `[${subtypeName} Markup - No comment text]`;
      }

      // Author
      let author = (annot.title ?? '').trim();
      if (!author) author = 'Unknown Reviewer';

      // Dates
      let createdDate = parsePdfDate(annot.creationDate ?? '');
      let modifiedDate = parsePdfDate(annot.modificationDate ?? '');
      if (!createdDate) createdDate = modifiedDate || nowFormatted();
      if (!modifiedDate) modifiedDate = createdDate;

      // Selected text — extract from highlighted/underlined regions
      let selectedText = '';
      if (TEXT_EXTRACT_TYPES.has(subtypeName) && annot.chars) {
        try {
          selectedText = annot.chars.map((c: { unicode: string }) => c.unicode).join('').trim();
        } catch {
          // not available
        }
      }

      // If pdf.js provides quadPoints (highlight quads), get text from content
      if (TEXT_EXTRACT_TYPES.has(subtypeName) && !selectedText && annot.rect) {
        try {
          const textContent = await page.getTextContent();
          const viewport = page.getViewport({ scale: 1 });
          const [, , , pageHeight] = [0, 0, viewport.width, viewport.height];
          const [ax0, ay0, ax1, ay1] = annot.rect;
          // PDF coordinate system: y=0 is bottom; text items use the same coords
          const words: string[] = [];
          for (const item of textContent.items as Array<{ str: string; transform: number[] }>) {
            if (!item.transform) continue;
            const tx = item.transform[4];
            const ty = item.transform[5];
            if (tx >= Math.min(ax0, ax1) - 2 &&
                tx <= Math.max(ax0, ax1) + 2 &&
                ty >= Math.min(ay0, ay1) - 5 &&
                ty <= Math.max(ay0, ay1) + 5) {
              words.push(item.str);
            }
          }
          selectedText = words.join(' ').trim();
          void pageHeight; // suppress unused warning
        } catch {
          // best-effort
        }
      }

      results.push({
        pdf_name: pdfName,
        pdf_url: pdfUrl,
        page_number: pageNum,
        annotation_type: subtypeName,
        comment_text: commentText,
        author,
        created_date: createdDate,
        modified_date: modifiedDate,
        rectangle_coordinates: coordsStr,
        selected_text: selectedText,
        assigned_to: 'Unassigned',
        priority: 'Medium',
        status: 'Pending',
        remarks: '',
        color: getAnnotationColor(annot.color),
      });
    }
  }

  return results;
}
